Use enum values from enumType in DiscriminatorColumn and check DiscriminatorMap values against it

whataboutpereira opened the pull request:
Fix for issue #11794.

Currently the values of enumType in DiscriminatorColumn are ignored which results in an error when generating a migration:

```
Doctrine\DBAL\Platforms\MySQL80Platform requires the values of a ENUM column to be specified.
```

Fixed by populating values from the enum and also prevent invalid entries in DiscriminatorMap.

whataboutpereira commented:
What's the correct way to turn off tests for DBAL versions not supporting enumType?

whataboutpereira commented:
MySQL also required. :)

greg0ire commented on src/Mapping/MappingException.php:
```diff
@@ -329,6 +329,24 @@ public static function fileMappingDriversRequireConfiguredDirectoryPath(string|n
+    public static function invalidEntriesInDiscriminatorMap(array $entries, string $owningClass, string $enumType): self
+    {
+        return new self(sprintf(
+            "The entries %s in the discriminator map of class '%s' do not correspond to enum cases of '%s'.",
```
You're wrapping the second `%s` with quotes, why not the first one as well? It might be empty.

whataboutpereira replied:
This exception is only thrown when there are actual non-matching values. Otherwise individual values are quoted when there's a mismatch.

greg0ire commented:
Retargeting to 3.6.x since this is a new feature.

derrabus commented on tests/Tests/ORM/Tools/SchemaToolTest.php:
```diff
@@ -271,6 +276,42 @@ public function testSetDiscriminatorColumnWithoutLength(): void
+        $metadata->setInheritanceType(ClassMetadata::INHERITANCE_TYPE_SINGLE_TABLE);
+        $metadata->setDiscriminatorColumn([
+            'name' => 'discriminator',
+            'type' => defined(Types::ENUM) ? 'enum' : 'string',
```
This constant must be defined at this point because you're skipping the test on older DBAL versions.

whataboutpereira replied:
Thanks.

derrabus commented on tests/Tests/ORM/Tools/SchemaToolTest.php:
```diff
@@ -271,6 +276,42 @@ public function testSetDiscriminatorColumnWithoutLength(): void
+        }
+
+        if (! $this->getTestEntityManager()->getConnection()->getDatabasePlatform() instanceof AbstractMySQLPlatform) {
+            self::markTestSkipped('Test valid for MySQL/MariaDB only.');
```
Why shouldn't we run this test on other DBMS?

whataboutpereira replied:
Tests were failing on SQLite. I'm not sure which other feature detection we should use to skip.

derrabus replied:
Why were they failing though?

whataboutpereira replied:
The more I look at it it the more confused I get. :) Now I'm thinking Types::ENUM doesn't need to be involved at all when setting the values in ClassMetadata, because you should still get the values from the enum class even if not using type enum in the database. But it's 2AM, will test tomorrow.

whataboutpereira commented:
> Can you please squash your commits together and give the remaining commit [a good message](https://www.doctrine-project.org/contribute/index.html#crafting-meaningful-commit-messages)?

Thanks, done now. Hopefully correctly. :)

whataboutpereira commented on src/Mapping/ClassMetadata.php:
```diff
@@ -2241,6 +2257,16 @@ public function setDiscriminatorMap(array $map): void
             );
         }
 
+        $values = $this->discriminatorColumn->options['values'] ?? null;
```
I'm now thinking we should wrap these checks in if ($this->discriminatorColumn?->enumType !== null), otherwise DiscriminatorColumn without enumType, but with options: ['values' => ['...']] will fail. Though values probably shouldn't be used without.

whataboutpereira committed:
Extract enum cases from enumType in DiscriminatorColumn
Check DiscriminatorMap keys match enum cases.
Test values are populated from enum cases and mismatched values throw an exception.
Fixes #11794
